### PR TITLE
Files plugin fixes/adaptation

### PR DESCRIPTION
**Pull request by Mingling94**
- Adapted to new API
- Fixed various bugs (Still addressing many concerns of #175 though)
- Share and shareascii for folders
- Also fixed some deprecated api usage (file.nickname -> file.siapath)
  - Reflect new `renew` attribute in files and file elements
- tools.notType(), function to log improper arguments
  - Check param in each file/folder function
- Removed file.size and file.count getters, trying not to cater file class to folder class. Folder should strictly inherit from file
- Name fields do not take up more space than needed

Note: Many buttons still not working. Much `TODO:`

### CODE REVIEW

File: plugins/Files/js/file.js [hunk 1]
```diff
@@ -98,42 +108,47 @@ var file = {
...
+	// Misc. functions
+	get parentFolders () {
+		// path of 'foo/bar/baz' would return folder objects whose paths are
+		// ['foo', 'foo/bar'] respectively
+		var parentFolders = [];
+		// iterate through parentFolder links to populate parentFolders
+		for (let i = this.parentFolder; i; i = i.parentFolder) {
+			parentFolders.push(i);
+		}
+		return parentFolders.reverse();
```
Comment (DavidVorick) [inline plugins/Files/js/file.js]: there's no builtin `split` function?

Comment (Mingling94) [inline plugins/Files/js/file.js]: This gets references to the folder objects, not the array of folder names (as the function `folderNames()` does)

File: plugins/Files/js/folder.js [hunk 2]
```diff
@@ -105,26 +77,85 @@ var folder = Object.assign(Object.create(file), {
 		tools.waterfall(functs, names, callback);
 	},
 
-	// Share .sia files into folder at destination with same structure
+	// Share .sia files of all contents (deep) to destination
```
Comment (DavidVorick) [inline plugins/Files/js/folder.js]: this wording is not really clear

File: plugins/Files/js/folder.js [hunk 3]
```diff
@@ -105,26 +77,85 @@ var folder = Object.assign(Object.create(file), {
...
+			qs: {
+				siapaths: this.paths,
+				destination: destination,
+			},
+		}, callback);
+	},
 
-		// Make folder at destination
-		mkdirp.sync(destination);
+	// Share ascii of all contents (deep)
```
Comment (DavidVorick) [inline plugins/Files/js/folder.js]: same thing. The word 'deep' really doesn't have any meaning

**Comment by DavidVorick** [comment]
You gotta fix the build so it doesn't fail randomly.

**Comment by Mingling94** [comment]
Oh nah this time it's because of linting errors that I neglected to check.

**Comment by Mingling94** [comment]
@DavidVorick @lukechampine I think I've addressed all of #175 now...

File: plugins/Files/js/browser.js [hunk 4]
```diff
@@ -226,34 +303,116 @@ var browser = {
...
+		if (itemCount === 0) {
+			tools.tooltip('No selected files', $('.controls .share').get(0));
+			return;
+		} else if (itemCount === 1) {
+			let file = files[0];
+			label = file.name;
+			paths = file.type === 'folder' ? file.paths : [file.path];
+		} else {
+			// Reduce array of files to array of all file paths
+			paths = files.reduce(function(a, b) {
```
Comment (lukechampine) [inline plugins/Files/js/browser.js]: reducing an array to an array sounds like a map, no? Or are the types different?

File: plugins/Files/js/browser.js [hunk 5]
```diff
@@ -226,34 +303,116 @@ var browser = {
...
 			return;
 		} else if (itemCount === 1) {
+			// Save file/folder into specific place
 			label = files[0].name;
+			destination = tools.dialog('save', {
+				title: 'Download ' + label,
+				defaultPath: label,
+			});
 		} else {
+			let totalCount = files.reduce(function(a, b) {
```
Comment (lukechampine) [inline plugins/Files/js/browser.js:401]: woo higher order functions

File: plugins/Files/js/file.js [hunk 6]
```diff
@@ -66,72 +83,86 @@ var file = {
 		}, callback);
 	},
 	share (destination, callback) {
-		if (typeof destination !== 'string' || destination.slice(-4) !== '.sia') {
-			console.error('Improper argument!', destination);
+		if (tools.notType(destination, 'string')) {
+			return;
+		} else if (destination.slice(-4) !== '.sia') {
+			console.error('Share path needs end in ".sia"!', destination);
```
Comment (lukechampine) [inline plugins/Files/js/file.js]: typo, but this is handled in the backend anyway

File: plugins/Files/js/folder.js [hunk 7]
```diff
@@ -48,103 +122,130 @@ var folder = Object.assign(Object.create(file), {
...
 
-		// Make corresponding array of file paths
-		names = names.map(name => `${destination}/${name}`);
+		var pathArray;
+		// File passed in is a file object
+		if (typeof f === 'object') {
+			pathArray = f.pathArray;
+		} else if (typeof f === 'string') {
+			// File passed in is a file path
+			pathArray = f.split('/');
```
Comment (lukechampine) [inline plugins/Files/js/folder.js:191]: I worry about the use of `/` as the path separator. Could break on Windows.

File: plugins/Files/js/fileElement.js [hunk 8]
```diff
@@ -6,97 +6,115 @@
...
-	} else if (f.uploadProgress < 100) {
-		timeText = f.uploadProgress.toFixed(0) + '%'; 
+	// Set detail text
+	var detailText;
+	if (f.uploadprogress === 0) {
+		detailText = 'Processing...';
+	} else if (f.uploadprogress < 100) {
+		detailText = f.uploadprogress.toFixed(0) + '%'; 
+	} else if (f.renewing) {
+		detailText = 'Renewing';
```
Comment (lukechampine) [inline plugins/Files/js/fileElement.js:35]: `Renewing` makes it sound like something is actively happening, when in reality the file is just set to be renewed as needed. I think in this case, nothing should be displayed at all.

**Pull request merged**